policy/release/repo_ids.rego

```diff
@@ -0,0 +1,170 @@
+#
+# METADATA
+# title: Repo IDs
+# description: >-
+#   Checks that all RPMs listed in the SBOM have a valid and known
+#   repository id.
+#
+package policy.release.repo_ids
+
+import rego.v1
+
+import data.lib
+
+# METADATA
+# title: Known repo id list provided
+# description: >-
+#   A list of valid and known repository ids should be available in the data.
+# custom:
+#   short_name: rpm_repo_ids_list_provided
+#   failure_msg: 'Valid repo id list not provided: %s'
+#   solution: >-
+#     Include a data source that provides a list of known repository ids, for example
+#     https://github.com/release-engineering/rhtap-ec-policy/blob/main/data/known_rpm_repositories.yml
+#   collections:
+#   - redhat
+#
+deny contains result if {
+	some problem in data_problems
+	result := lib.result_helper(rego.metadata.chain(), [problem])
+}
+
+# METADATA
+# title: All rpms have known repo ids
+# description: >-
+#   The list of rpm packages in the SBOM will be inspected. A violation will be produced if any
+#   of them do not specify a "repository_id" from the list of known accepted repository ids.
+# custom:
+#   short_name: rpm_repo_ids_valid
+#   failure_msg: 'RPM repo id check failed: %s'
+#   solution: >-
+#     Ensure every rpm is from a known and accepted repository and that the data in the SBOM
+#     correctly records that.
+#   # Todo: Until the sbom generation is upated, this will always fail,
+#   # so for now we don't include it in the redhat collection.
+#   # See https://issues.redhat.com/browse/STONEBLD-2638
+#   #collections:
+#   #- redhat
+#
+deny contains result if {
+	# Don't bother with this unless there's some data available
+	count(data_problems) == 0
+
+	some problem in repo_id_problems
+	result := lib.result_helper(rego.metadata.chain(), [problem])
+}
+
+data_problems contains reason if {
+	not known_repo_ids
+	reason := "The 'known_rpm_repositories' key was not found in the data"
+}
+
+data_problems contains reason if {
+	not is_array(known_repo_ids)
+	reason := "The 'known_rpm_repositories' data is not an array"
+}
+
+data_problems contains reason if {
+	count(known_repo_ids) == 0
+	reason := "The 'known_rpm_repositories' data is empty"
+}
+
+data_problems contains reason if {
+	is_array(known_repo_ids)
+	non_strings := [r |
+		some r in known_repo_ids
+		not is_string(r)
+	]
+	count(non_strings) > 0
+	reason := "The 'known_rpm_repositories' data is not a list of strings"
+}
```

Comment:
Usually we do this via a JSON schema check, no objection to doing it this way, just an observation.

Reply:
+1 to using a schema. You can also define that the list must be non-empty which should simplify the checks here.

Reply:
Done in latest revision.